provider/qsd.py

```diff
@@ -0,0 +1,446 @@
+    cmd = ""
+    for dev in devs:
+        cmd += " %s" % dev.cmdline()
+    if cmd:
+        machine_type = params.get("machine_type", "")
+        if machine_type.startswith("q35"):
+            busid = "pcie_vubp_root_port_%d" % addr
+            bus = "-device pcie-root-port,id=%s,bus=pcie.0,addr=%d " % (busid, addr)
+            cmd = bus + cmd + ",bus=%s" % busid
+        params["extra_params"] = cmd
+        LOG_JOB.info("Ready add %s into VM command line", cmd)
+        return cmd
+
+
+def get_qsd_name_by_image(img_name, params):
+    """Get QSD name by image """
+    if params.get("drive_format_%s" % img_name) != "vhost-user-blk-pci":
+        return
+    qsd_name = ""
+
+    for qsd in params["qsd_namespaces"].split():
+        qsd_imgs = params.object_params(qsd).get("qsd_images", "").split()
+        if img_name in qsd_imgs:
+            qsd_name = qsd
+            break
+    return qsd_name
+
+
+def create_vubp_devices(vm, img_name, params, bus=None):
+    """Create vhost-user-blk-pci relevant devices"""
+    qsd_name = get_qsd_name_by_image(img_name, params)
+    if not qsd_name:
+        raise ValueError("Can not find QSD")
+
+    qsd_params = params.object_params(qsd_name)
+    qsd_basedir = qsd_params.get("qsd_basedir", data_dir.get_data_dir() + "/qsd/%s" % qsd_name)
+
+    qid = "qsd_%s" % qsd_name
+    # qbus_id = "qsd_bus_%s" % qsd_name
+    img_sock = "%s/%s_vhost_user_%s.sock" % (qsd_basedir, qsd_name, img_name)
+    devices = []
+    if vm and not vm.devices.get_by_qid(qid):
+        # Declare virtual QSD daemon
+        qsd = qdevices.QDaemonDev("qsd", aobject=qsd_name, child_bus=qdevices.QUnixSocketBus(img_sock, qsd_name))
+        vm.devices.insert(qsd)
+
+    machine_type = params.get("machine_type", "")
+    qbus_type = "PCI"
+    if machine_type.startswith("q35"):
+        qbus_type = "PCIE"
+
+    char_params = Params()
+    char_params["backend"] = "socket"
+    char_params["id"] = 'char_qsd_%s' % qsd_name
+    char_params["path"] = img_sock
+    sock_bus = {'busid': img_sock}
+    char = qdevices.CharDevice(char_params, parent_bus=sock_bus)
+    char.set_param("server", "off")
+    char.set_aid(char_params["id"])
+    devices.append(char)
+
+    qdriver = "vhost-user-blk-pci"
+
+    if bus is None:
+        bus = {"type": qbus_type}
+
+    dev_params = {"id": "vubp_%s" % img_name,
+                  "chardev": char.get_qid()
+                  }
+    vubp_driver_props = json.loads(params.get("image_vubp_props_%s" % img_name, "{}"))
+    dev_params.update(vubp_driver_props)
+    vubp = qdevices.QDevice(qdriver, params=dev_params, parent_bus=bus)
+    vubp.set_aid(dev_params["id"])
+    devices.append(vubp)
+    LOG_JOB.info("create_vubp_devices %s %s", img_name, devices)
+    return devices
+
+
+def insert_vubp_devices(vm, img_name, params, bus=None):
+    """Insert vhost-user-blk-pci into VM internal devices"""
+    devs = create_vubp_devices(vm, img_name, params, bus)
+    vm.devices.insert(devs)
```

Comment:
Not sure, why we need this function? When and how can we call it?

We've already have vm.devices.insert(qsd) in create_vubp_devices.

Reply:
you are right,will be removed